File: rust/src/step8_macros.rs
// support precompiled regexes in reader.rs
#![feature(phase)]
#[phase(plugin)]
extern crate regex_macros;
extern crate regex;

use std::os;

use types::{MalVal,MalRet,MalFunc,
            Nil,False,Sym,List,Vector,Func,
            _nil,symbol,string,list,malfunc,malfuncd};
use env::{Env,env_new,env_bind,env_root,env_find,env_set,env_get};
mod readline;
mod types;
mod reader;
mod printer;
mod env;
mod core;

// read
fn read(str: String) -> MalRet {
    reader::read_str(str)
}

// eval
fn is_pair(x: MalVal) -> bool {
    match *x {
        List(ref lst) => lst.len() > 0,
        _ => false,
    }
}

fn quasiquote(ast: MalVal) -> MalVal {
    if !is_pair(ast.clone()) {
        return list(vec![symbol("quote"), ast])
    }

    match *ast.clone() {
        List(ref args) => {
            let ref a0 = args[0];
            match **a0 {
                Sym(ref s) => {
                    if s.to_string() == "unquote".to_string() {
                        let ref a1 = args[1];
                        return a1.clone();
                    }
                },
                _ => (),
            }
            if is_pair(a0.clone()) {
                match **a0 {
                    List(ref a0args) => {
                        let a00 = a0args[0].clone();
                        match *a00 {
                            Sym(ref s) => {
                                if s.to_string() == "splice-unquote".to_string() {
                                    return list(vec![symbol("concat"),
                                                     a0args[1].clone(),
                                                     quasiquote(list(args.slice(1,args.len()).to_vec()))])
                                }
                            },
                            _ => (),
                        }
                    },
                    _ => (),
                }
            }
            let rest = list(args.slice(1,args.len()).to_vec());
            return list(vec![symbol("cons"),
                             quasiquote(a0.clone()),
                             quasiquote(rest)])
        },
        _ => _nil(), // should never reach
    }
}

fn is_macro_call(ast: MalVal, env: Env) -> bool {
    match *ast {
        List(ref lst) => {
            let ref a0 = *lst[0];
            match *a0 {
                Sym(ref a0sym) => {
                    if env_find(env.clone(), a0sym.to_string()).is_some() {
                        match env_get(env, a0sym.to_string()) {
                            Ok(f) => {
                                match *f {
                                    MalFunc(ref mfd) => {
                                        mfd.is_macro
                                    },
                                    _ => false,
                                }
                            },
                            _ => false,
                        }
                    } else {
                        false
                    }
                },
                _ => false,
            }
        },
        _ => false,
    }
}

fn macroexpand(mut ast: MalVal, env: Env) -> MalRet {
    while is_macro_call(ast.clone(), env.clone()) {
        match *ast.clone() {
            List(ref args) => {
                let ref a0 = args[0];
                match **a0 {
                    Sym(ref s) => {
                        match env_get(env.clone(), s.to_string()) {
                            Ok(mf) => {
                                match *mf {
                                    MalFunc(_) => {
                                        match mf.apply(args.slice(1,args.len()).to_vec()) {
                                            Ok(r) => ast = r,
                                            Err(e) => return Err(e),
                                        }
                                    },
                                    _ => break,
                                }
                            },
                            Err(e) => return Err(e),
                        }
                    },
                    _ => break,
                }
            },
            _ => break,
        }
    }
    Ok(ast)
}

fn eval_ast(ast: MalVal, env: Env) -> MalRet {
    let ast2 = ast.clone();
    match *ast2 {
    //match *ast {
        Sym(ref sym) => {
            env_get(env.clone(), sym.clone())
        },
        List(ref a) => {
            let mut ast_vec : Vec<MalVal> = vec![];
            for mv in a.iter() {
                let mv2 = mv.clone();
                match eval(mv2, env.clone()) {
                    Ok(mv) => { ast_vec.push(mv); },
                    Err(e) => { return Err(e); },
                }
            }
            Ok(list(ast_vec))
        },
        _ => {
            Ok(ast)
        }
    }
}

fn eval(mut ast: MalVal, mut env: Env) -> MalRet {
    'tco: loop {

    //println!("eval: {}, {}", ast, env.borrow());
    //println!("eval: {}", ast);
    let mut ast2 = ast.clone();
    match *ast2 {
        List(_) => (),  // continue
        _ => return eval_ast(ast2, env),
    }

    // apply list
    match macroexpand(ast2, env.clone()) {
        Ok(a) => {
            ast2 = a;
        },
        Err(e) => return Err(e),
    }
    match *ast2 {
        List(_) => (),  // continue
        _ => return Ok(ast2),
    }
    let ast3 = ast2.clone();

    match *ast2 {
        List(ref args) => {
            if args.len() == 0 { 
                return Ok(ast3);
            }
            let ref a0 = *args[0];
            match *a0 {
                Sym(ref a0sym) => {
                    match a0sym.as_slice() {
                        "def!" => {
                            let a1 = (*args)[1].clone();
                            let a2 = (*args)[2].clone();
                            let res = eval(a2, env.clone());
                            match res {
                                Ok(r) => {
                                    match *a1 {
                                        Sym(ref s) => {
                                            env_set(&env.clone(), s.clone(), r.clone());
                                            return Ok(r);
                                        },
                                        _ => {
                                            return Err("def! of non-symbol".to_string())
                                        }
                                    }
                                },
                                Err(e) => return Err(e),
                            }
                        },
                        "let*" => {
                            let let_env = env_new(Some(env.clone()));
                            let a1 = (*args)[1].clone();
                            let a2 = (*args)[2].clone();
                            match *a1 {
                                List(ref binds) | Vector(ref binds) => {
                                    let mut it = binds.iter();
                                    while it.len() >= 2 {
                                        let b = it.next().unwrap();
                                        let exp = it.next().unwrap();
                                        match **b {
                                            Sym(ref bstr) => {
                                                match eval(exp.clone(), let_env.clone()) {
                                                    Ok(r) => {
                                                        env_set(&let_env, bstr.clone(), r);
                                                    },
                                                    Err(e) => {
                                                        return Err(e);
                                                    },
                                                }
                                            },
                                            _ => {
                                                return Err("let* with non-symbol binding".to_string());
                                            },
                                        }
                                    }
                                },
                                _ => return Err("let* with non-list bindings".to_string()),
                            }
                            ast = a2;
                            env = let_env.clone();
                            continue 'tco;
                        },
                        "quote" => {
                            return Ok((*args)[1].clone());
                        },
                        "quasiquote" => {
                            let a1 = (*args)[1].clone();
                            ast = quasiquote(a1);
                            continue 'tco;
                        },
                        "defmacro!" => {
                            let a1 = (*args)[1].clone();
                            let a2 = (*args)[2].clone();
                            match eval(a2, env.clone()) {
                                Ok(r) => {
                                    match *r {
                                        MalFunc(ref mfd) => {
                                            match *a1 {
                                                Sym(ref s) => {
                                                    let mut new_mfd = mfd.clone();
                                                    new_mfd.is_macro = true;
                                                    let mf = malfuncd(new_mfd);
                                                    env_set(&env.clone(), s.clone(), mf.clone());
                                                    return Ok(mf);
                                                },
                                                _ => return Err("def! of non-symbol".to_string()),
                                            }
                                        },
                                        _ => return Err("def! of non-symbol".to_string()),
                                    }
                                },
                                Err(e) => return Err(e),
                            }
                        },
                        "macroexpand" => {
                            let a1 = (*args)[1].clone();
                            return macroexpand(a1, env.clone())
                        },
                        "do" => {
                            let el = list(args.slice(1,args.len()-1).to_vec());
                            match eval_ast(el, env.clone()) {
                                Err(e) => return Err(e),
                                Ok(_) => {
                                    let ref last = args[args.len()-1];
                                    ast = last.clone();
                                    continue 'tco;
                                },
                            }
                        },
                        "if" => {
                            let a1 = (*args)[1].clone();
                            let cond = eval(a1, env.clone());
                            if cond.is_err() { return cond; }
                            match *cond.unwrap() {
                                False | Nil => {
                                    if args.len() >= 4 {
                                        let a3 = (*args)[3].clone();
                                        ast = a3;
                                        env = env.clone();
                                        continue 'tco;
                                    } else {
                                        return Ok(_nil());
                                    }
                                },
                                _ => {
                                    let a2 = (*args)[2].clone();
                                    ast = a2;
                                    env = env.clone();
                                    continue 'tco;
                                },
                            }
                        },
                        "fn*" => {
                            let a1 = (*args)[1].clone();
                            let a2 = (*args)[2].clone();
                            return Ok(malfunc(eval, a2, env.clone(), a1));
                        },
                        "eval" => {
                            let a1 = (*args)[1].clone();
                            match eval(a1, env.clone()) {
                                Ok(exp) => {
                                    ast = exp;
                                    env = env_root(&env);
                                    continue 'tco;
                                },
                                Err(e) => return Err(e),
                            }
                        },
                        _ => ()
                    }
                }
                _ => (),
            }
            // function call
            /*
            if is_macro_call(ast3.clone(), env.clone()) {
                println!("macro call");
            }
            */
            return match eval_ast(ast3, env.clone()) {
                Err(e) => Err(e),
                Ok(el) => {
                    match *el {
                        List(ref args) => {
                            let args2 = args.clone();
                            match *args2[0] {
                                Func(f) => f(args.slice(1,args.len()).to_vec()),
                                MalFunc(ref mf) => {
                                    let mfc = mf.clone();
                                    let alst = list(args.slice(1,args.len()).to_vec());
                                    let new_env = env_new(Some(mfc.env.clone()));
                                    match env_bind(&new_env, mfc.params, alst) {
                                        Ok(_) => {
                                            ast = mfc.exp;
                                            env = new_env;
                                            continue 'tco;
                                        },
                                        Err(e) => Err(e),
                                    }
                                },
                                _ => Err("attempt to call non-function".to_string()),
                            }
                        }
                        _ => Err("Invalid apply".to_string()),
                    }
                }
            }
        }
        _ => return Err("Expected list".to_string()),
    }

    }
}

// print
fn print(exp: MalVal) -> String {
    exp.pr_str(true)
}

fn rep(str: String, env: Env) -> Result<String,String> {
    match read(str) {
        Err(e) => Err(e),
        Ok(ast) => {
            //println!("read: {}", ast);
            match eval(ast, env) {
                Err(e)  => Err(e),
                Ok(exp) => Ok(print(exp)),
            }
        }
    }
}

fn main() {
    // core.rs: defined using rust
    let repl_env = env_new(None);
    for (k, v) in core::ns().into_iter() { env_set(&repl_env, k, v); }
    // see eval() for definition of "eval"
    env_set(&repl_env, "*ARGV*".to_string(), list(vec![]));

    // core.mal: defined using the language itself
    let _ = rep("(def! not (fn* (a) (if a false true)))".to_string(),
                repl_env.clone());
    let _ = rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))".to_string(),
                repl_env.clone());

    // Invoked with command line arguments
    let args = os::args();
    if args.len() > 1 {
        let mv_args = args.slice(2,args.len()).iter()
            .map(|a| string(a.to_string()))
            .collect::<Vec<MalVal>>();
        env_set(&repl_env, "*ARGV*".to_string(), list(mv_args));
        match rep("(load-file \"".to_string() + args[1] + "\")".to_string(),
                  repl_env.clone()) {
            Ok(_) => {
                os::set_exit_status(0);
                return;
            },
            Err(str) => {
                println!("Error: {}", str);
                os::set_exit_status(1);
                return;
            },
        }
    }

    loop {
        let line = readline::mal_readline("user> ");
        match line { None => break, _ => () }
        match rep(line.unwrap(), repl_env.clone()) {
            Ok(str)  => println!("{}", str),
            Err(str) => println!("Error: {}", str),
        }
    }
}

File: nim/step8_macros.nim
import rdstdin, tables, sequtils, os, types, reader, printer, env, core

proc read(str: string): MalType = str.read_str

proc is_pair(x: MalType): bool =
  x.kind in {List, Vector} and x.list.len > 0

proc quasiquote(ast: MalType): MalType =
  if not ast.is_pair:
    return list(symbol "quote", ast)
  elif ast.list[0] == symbol "unquote":
    return ast.list[1]
  elif ast.list[0].is_pair and ast.list[0].list[0] == symbol "splice-unquote":
    return list(symbol "concat", ast.list[0].list[1],
      quasiquote(list ast.list[1 .. ^1]))
  else:
    return list(symbol "cons", quasiquote(ast.list[0]), quasiquote(list(ast.list[1 .. ^1])))

proc is_macro_call(ast: MalType, env: Env): bool =
  ast.kind == List and ast.list[0].kind == Symbol and
    env.find(ast.list[0].str) != nil and env.get(ast.list[0].str).macro_q

proc macroexpand(ast: MalType, env: Env): MalType =
  result = ast
  while result.is_macro_call(env):
    let mac = env.get(result.list[0].str)
    result = mac.malfun.fn(result.list[1 .. ^1]).macroexpand(env)

proc eval(ast: MalType, env: Env): MalType

proc eval_ast(ast: MalType, env: var Env): MalType =
  case ast.kind
  of Symbol:
    result = env.get(ast.str)
  of List:
    result = list ast.list.mapIt(MalType, it.eval(env))
  of Vector:
    result = vector ast.list.mapIt(MalType, it.eval(env))
  of HashMap:
    result = hash_map()
    for k, v in ast.hash_map.pairs:
      result.hash_map[k] = v.eval(env)
  else:
    result = ast

proc eval(ast: MalType, env: Env): MalType =
  var ast = ast
  var env = env

  template defaultApply =
    let el = ast.eval_ast(env)
    let f = el.list[0]
    case f.kind
    of MalFun:
      ast = f.malfun.ast
      env = initEnv(f.malfun.env, f.malfun.params, list(el.list[1 .. ^1]))
    else:
      return f.fun(el.list[1 .. ^1])

  while true:
    if ast.kind != List: return ast.eval_ast(env)

    ast = ast.macroexpand(env)
    if ast.kind != List: return ast
    if ast.list.len == 0: return ast

    let a0 = ast.list[0]
    case a0.kind
    of Symbol:
      case a0.str
      of "def!":
        let
          a1 = ast.list[1]
          a2 = ast.list[2]
        return env.set(a1.str, a2.eval(env))

      of "let*":
        let
          a1 = ast.list[1]
          a2 = ast.list[2]
        var let_env = Env(env)
        case a1.kind
        of List, Vector:
          for i in countup(0, a1.list.high, 2):
            let_env.set(a1.list[i].str, a1.list[i+1].eval(let_env))
        else: raise newException(ValueError, "Illegal kind in let*")
        ast = a2
        env = let_env
        # Continue loop (TCO)

      of "quote":
        return ast.list[1]

      of "quasiquote":
        ast = ast.list[1].quasiquote
        # Continue loop (TCO)

      of "defmacro!":
        var fun = ast.list[2].eval(env)
        fun.malfun.is_macro = true
        return env.set(ast.list[1].str, fun)

      of "macroexpand":
        return ast.list[1].macroexpand(env)

      of "do":
        let last = ast.list.high
        let el = (list ast.list[1 .. <last]).eval_ast(env)
        ast = ast.list[last]
        # Continue loop (TCO)

      of "if":
        let
          a1 = ast.list[1]
          a2 = ast.list[2]
          cond = a1.eval(env)

        if cond.kind in {Nil, False}:
          if ast.list.len > 3: ast = ast.list[3]
          else: ast = nilObj
        else: ast = a2

      of "fn*":
        let
          a1 = ast.list[1]
          a2 = ast.list[2]
        var env2 = env
        let fn = proc(a: varargs[MalType]): MalType =
          var newEnv = initEnv(env2, a1, list(a))
          a2.eval(newEnv)
        return malfun(fn, a2, a1, env)

      else:
        defaultApply()

    else:
      defaultApply()

proc print(exp: MalType): string = exp.pr_str

var repl_env = initEnv()

for k, v in ns.items:
  repl_env.set(k, v)
repl_env.set("eval", fun(proc(xs: varargs[MalType]): MalType = eval(xs[0], repl_env)))
var ps = commandLineParams()
repl_env.set("*ARGV*", list((if paramCount() > 1: ps[1..ps.high] else: @[]).map(str)))


# core.nim: defined using nim
proc rep(str: string): string {.discardable.} =
  str.read.eval(repl_env).print

# core.mal: defined using mal itself
rep "(def! not (fn* (a) (if a false true)))"
rep "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))"
rep "(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))"
rep "(defmacro! or (fn* (& xs) (if (empty? xs) nil (if (= 1 (count xs)) (first xs) `(let* (or_FIXME ~(first xs)) (if or_FIXME or_FIXME (or ~@(rest xs))))))))"

if paramCount() >= 1:
  rep "(load-file \"" & paramStr(1) & "\")"
  quit()

while true:
  try:
    let line = readLineFromStdin("user> ")
    echo line.rep
  except Blank: discard
  except:
    echo getCurrentExceptionMsg()
    echo getCurrentException().getStackTrace()
